cloudsync/smartsync.py

```diff
@@ -0,0 +1,192 @@
+        #       or maybe it does make sense to add them to the requestset? that way they will get subscribed too
+        #       or maybe always sync all folders at all times, and only files are smartsynced
+        ents = self.lookup_path(REMOTE, remote_path)
+        for ent in ents:
+            self.requestset.add(ent)
+        return ents
+
+    def smart_unsync_ent(self, ent):
+        self.requestset.discard(ent)
+        self.excludeset.add(ent)
+
+    def smart_unsync_path(self, remote_path):
+        for ent in tuple(self.requestset):
+            if ent[REMOTE].path == remote_path:
+                self.state.smart_unsync_ent(ent)
+                break
+
+    def smart_listdir(self, local_path, remote_path):
+        """Lists all files in a remote folder by looking in the state db, and doesn't hit the provider api."""
+        prov = self.providers[REMOTE]
+        for ent, _ignored_rel_path in self.get_kids(remote_path, REMOTE):
+            remote_ent_path = ent[REMOTE].path
+            _ignored_remote_ent_parent, remote_ent_name = self.providers[REMOTE].split(ent[REMOTE].path)
+            if prov.paths_match(prov.dirname(remote_ent_path), remote_path):
+                entry = ent[REMOTE]
+                retval = OInfo(otype=entry.otype,
+                                 oid=entry.oid,
+                                 hash=None,
+                                 path=self.providers[LOCAL].join(local_path, remote_ent_name),
+                                 size=entry.size,
+                                 name=remote_ent_name,
+                                 mtime=entry.mtime,
+                                 shared=False,  # TODO: ent[REMOTE].shared,
+                                 readonly=False,  # TODO: ent[REMOTE].readonly
+                                 )
+                yield retval  # TODO: this needs to yield DirInfo, also translate back to local
+
+    @property
+    def changes(self):
+        # Filter the changes to just those that are in the requestset, or those whose get_latest is stale viz. change_time
+        changes = super().changes
+        for ent in self.requestset:
+            if ent[LOCAL].path in changes:
+                yield ent
+
+    @property
+    def _changeset(self):
+        folders = [ent for ent in self._changeset_storage if ent[REMOTE].otype == DIRECTORY or 
+                   ent[REMOTE].changed > ent[REMOTE]._last_gotten]  # pylint: disable=protected-access
+        retval = self.requestset.intersection(self._changeset_storage).union(folders).difference(self.excludeset)
+        return retval
+
+    # def change(self, age):
+    #     super().change(age)
+    #     requested_changes = self.requestset.intersection(self._changeset):
+    #     if requested_changes:
+
+    # def unconditionally_get_latest(self, ent, i):
+    #     # TODO: get enhanced data (mtime, size, etc)
+    #     super().unconditionally_get_latest(ent, i)
+
+
+class SmartCloudSync(CloudSync):
```

Comment:
I think in a few places in the fs (opening a handle, stat, get_security_by_name) we are going to want to ask smartsync for the OInfo for a local path that may be smartsynced (or None if missing). I think we need a `smart_info_path` function here unless there is a better way to retrieve this information from the state table.

Reply:
How is what you are requesting different from a regular call to `info_path`?

Reply:
Calling the remote providers `info_path` is an online operation that requires hitting the api for most if not all providers. We want to just look at what's in the statedb.

Reply:
Do you want the OInfo for a file that has not been synced?

Reply:
Yes, we want the Oinfo for a file hasn't been synced down locally yet but exists in the state db. For instance, if something shows up in the `smart_listdir`, this method should also return the OInfo for this ent if queried directly by path.

Reply:
done